examples/speech_recognition/modules/specaugment.py

```diff
@@ -0,0 +1,75 @@
+"""SpecAugment Implementation for Pytorch.
+Related paper : https://arxiv.org/pdf/1904.08779.pdf
+"""
+
+import torch
+import torch.nn as nn
+import numpy as np
+import random
+
+class SpecAugment(nn.Module):
+    """
+    Class that handles 'SpecAugment'. It stores the parameters and calls the function according to the probabilistic rate.
+    """
+
+    def __init__(self, frequency_masking_pars, time_masking_pars,
+                 frequency_masking_num, time_masking_num, rate=1.0):
+        super().__init__()
+        self.frequency_masking_pars = frequency_masking_pars
+        self.time_masking_pars = time_masking_pars
+        self.frequency_masking_num = frequency_masking_num
+        self.time_masking_num = time_masking_num
+        self.rate = rate
+
+    def forward(self, batch):
+        """
+        Performs 'SpecAugment' on a batch
+        :param batch: A batch as returned by the 'collater' method
+        :return: The same batch with the masking applied by 'SpecAugment'
+        """
+        new_spectrograms = []
+        x = batch['net_input']['src_tokens']
+        for spectrogram in x:
+            if random.random() < self.rate:
+                sample = specaugment(spectrogram, self.frequency_masking_pars,
+                                    self.time_masking_pars, self.frequency_masking_num, self.time_masking_num,
+                                     )
+            else:
+                sample = spectrogram
+            new_spectrograms += [sample]
+
+        new_spectrograms = torch.stack(new_spectrograms)
+        batch['net_input']['src_tokens'] = new_spectrograms
+        return batch
+
+
+def specaugment(mel_spectrogram, frequency_masking_para=27,
+                 time_masking_para=100, frequency_masking_num=1, time_masking_num=1):
+    """
+    Spec augmentation Calculation Function.
+    'SpecAugment' has 3 steps for audio data augmentation, but only 2 are implemented in this version.
+    First step is frequency masking, second step is time masking.
+    :param mel_spectrogram: 2-dimensional Tensor representing a spectrogram
```

Comment:
in this repo the convention/format used is comments is:
```
Args:
  ....
```
Moreover, may you please also add a description of what each of the 2 dimentions mean? Thanks,

Reply:
I hope that it's better now